File: 정민선/프로그래머스/[PCCE 기출문제] 10번 공원.py
def solution(mats, park):
    answer = -1
    n = len(park)
    m = len(park[0])
    mats.sort(reverse=True)
    
    def inrange(x, y):
        return 0 <= x < n and 0 <= y < m
    
    def check(x, y, mat):
        if not inrange(x+mat-1, y+mat-1):
            return False
        
        for a in range(x, x+mat):
            for b in range(y, y+mat):
                if park[a][b] != "-1":
                    return False
        return True
    
    for mat in mats:
        for i in range(n-mat+1):
            for j in range(m-mat+1):
                if park[i][j] == "-1":
                    if check(i, j, mat):
                        answer = mat    
                        return answer
                    
    return answer

mats = [5,3,2]
park = [["A", "A", "-1", "B", "B", "B", "B", "-1"], ["A", "A", "-1", "B", "B", "B", "B", "-1"], ["-1", "-1", "-1", "-1", "-1", "-1", "-1", "-1"], ["D", "D", "-1", "-1", "-1", "-1", "E", "-1"], ["D", "D", "-1", "-1", "-1", "-1", "-1", "F"], ["D", "D", "-1", "-1", "-1", "-1", "E", "-1"]]

answer = solution(mats, park)
print(answer)

File: 이진영/프로그래머스/[PCCE 기출문제] 10번 공원.java
import java.util.*;

class Solution {
    public int solution(int[] mats, String[][] park) {
        int r = park.length;
        int c = park[0].length;
        Arrays.sort(mats);
        
        int idx = -1;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (!park[i][j].equals("-1")) continue;
                
                for (int mat = idx != -1 ? idx : 0; mat < mats.length; mat++) {
                    boolean flag = false;
                    for (int i2 = i; i2 < i + mats[mat]; i2++) {
                        if (i2 >= r) {
                            flag = true;
                            break;
                        }
                        for (int j2 = j; j2 < j + mats[mat]; j2++) {
                            if (j2 >= c || !park[i2][j2].equals("-1")) {
                                flag = true;
                                break;
                            }
                        }
                        if (flag) break;
                    }
                    if (flag) break;
                    else idx = mat;
                }
                
                if (idx == mats.length - 1) {
                    return mats[idx];
                }
            }
        }
        
        return idx != -1 ? mats[idx] : -1;
    }
}